realtek: add support for TP-Link SG2210P

mrnuke opened the pull request:
Add support for the TP-Link SG2210P switch. This is an RTL8380 based
switch with eight RJ-45 ports with 802.3af PoE, and two SFP ports.

This device shares the same board with the SG2008P and SG2008. To
model this, declare all the capabilities in the sg2xxx dtsi, and
disable unpopulated on the lower end models.

Specifications:
---------------
 - SoC:       Realtek RTL8380M
 - Flash:     32 MiB SPI flash (Vendor varies)
 - RAM:	      256 MiB (Vendor varies)
 - Ethernet:  8x 10/100/1000 Mbps with PoE (all ports)
              2x SFP ports
 - Buttons:   1x "Reset" button on front panel
 - Power:     53.5V DC barrel jack
 - UART:      1x serial header, unpopulated
 - PoE:       2x TI TPS23861 I2C PoE controller

Works:
------
  - (8) RJ-45 ethernet ports
  - (2) SFP ports (with caveats)
  - Switch functions
  - System LED

Not yet enabled:
----------------
  - Power-over-Ethernet (driver works, but doesn't enable "auto" mode)
  - PoE LEDs

Enabling SFP ports:
-------------------

The SFP port control lines are hardwired, except for tx-disable. These
lines are controller by the RTL8231 in shift register mode. There is
no driver support for this yet.

However, to enable the lasers on SFP1 and SFP2 respectively:

    echo 0x0510ff00 > /sys/kernel/debug/rtl838x/led/led_p_en_ctrl
    echo      0x140 > /sys/kernel/debug/rtl838x/led/led_sw_p_ctrl.26
    echo      0x140 > /sys/kernel/debug/rtl838x/led/led_sw_p_ctrl.24

Install via serial console/tftp:
--------------------------------

The footprints R27 (0201) and R28 (0402) are not populated. To enable
serial console, 50 ohm resistors should be soldered -- any value from
0 ohm to 50 ohm will work. R27 can be replaced by a solder bridge.

The u-boot firmware drops to a TP-Link specific "BOOTUTIL" shell at
38400 baud. There is no known way to exit out of this shell, and no
way to do anything useful.

Ideally, one would trick the bootloader into flashing the sysupgrade
image first. However, if the image exceeds 6MiB in size, it will not
work. The sysupgrade image can also be flashed. To install OpenWRT:

Prepare a tftp server with:
 1. server address: 192.168.0.146
 2. the image as: "uImage.img"

Power on device, and stop boot by pressing any key.
Once the shell is active:
 1. Ground out the CLK (pin 16) of the ROM (U7)
 2. Select option "3. Start"
 3. Bootloader notes that "The kernel has been damaged!"
 4. Release CLK as sson as bootloader thinks image is corrupted.
 5. Bootloader enters automatic recovery -- details printed on console
 6. Watch as the bootloader flashes and boots OpenWRT.

mrnuke committed:
realtek: add support for TP-Link SG2210P

Add support for the TP-Link SG2210P switch. This is an RTL8380 based
switch with eight RJ-45 ports with 802.3af PoE, and two SFP ports.

This device shares the same board with the SG2008P and SG2008. To
model this, declare all the capabilities in the sg2xxx dtsi, and
disable unpopulated on the lower end models.

Specifications:
---------------
 - SoC:       Realtek RTL8380M
 - Flash:     32 MiB SPI flash (Vendor varies)
 - RAM:	      256 MiB (Vendor varies)
 - Ethernet:  8x 10/100/1000 Mbps with PoE (all ports)
              2x SFP ports
 - Buttons:   1x "Reset" button on front panel
 - Power:     53.5V DC barrel jack
 - UART:      1x serial header, unpopulated
 - PoE:       2x TI TPS23861 I2C PoE controller

Works:
------
  - (8) RJ-45 ethernet ports
  - (2) SFP ports (with caveats)
  - Switch functions
  - System LED

Not yet enabled:
----------------
  - Power-over-Ethernet (driver works, but doesn't enable "auto" mode)
  - PoE LEDs

Enabling SFP ports:
-------------------

The SFP port control lines are hardwired, except for tx-disable. These
lines are controller by the RTL8231 in shift register mode. There is
no driver support for this yet.

However, to enable the lasers on SFP1 and SFP2 respectively:

    echo 0x0510ff00 > /sys/kernel/debug/rtl838x/led/led_p_en_ctrl
    echo      0x140 > /sys/kernel/debug/rtl838x/led/led_sw_p_ctrl.26
    echo      0x140 > /sys/kernel/debug/rtl838x/led/led_sw_p_ctrl.24

Install via serial console/tftp:
--------------------------------

The footprints R27 (0201) and R28 (0402) are not populated. To enable
serial console, 50 ohm resistors should be soldered -- any value from
0 ohm to 50 ohm will work. R27 can be replaced by a solder bridge.

The u-boot firmware drops to a TP-Link specific "BOOTUTIL" shell at
38400 baud. There is no known way to exit out of this shell, and no
way to do anything useful.

Ideally, one would trick the bootloader into flashing the sysupgrade
image first. However, if the image exceeds 6MiB in size, it will not
work. The sysupgrade image can also be flashed. To install OpenWRT:

Prepare a tftp server with:
 1. server address: 192.168.0.146
 2. the image as: "uImage.img"

Power on device, and stop boot by pressing any key.
Once the shell is active:
 1. Ground out the CLK (pin 16) of the ROM (U7)
 2. Select option "3. Start"
 3. Bootloader notes that "The kernel has been damaged!"
 4. Release CLK as sson as bootloader thinks image is corrupted.
 5. Bootloader enters automatic recovery -- details printed on console
 6. Watch as the bootloader flashes and boots OpenWRT.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
realtek: split TP-Link SG2000 series devicetree

The TP-Link TL-SG2008, TL-SG2008P, and TL-SG2210P use the same board.
The main difference is that some footprints are not populated in the
lower-end models. To model this with minimal duplication, move the
devicetree to a common dtsi, leaving out just the board name.

Use this opportunity to fix the labeling of ethernet ports. This
series has its ethernet ports in the rear. They are labeled backwards.
The correct port labeling is lan8->lan1. To resolve this, fix the phy
mapping in the "ports" node.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
realtek: tl-sg2xxx: read MAC address from nvmem-cells

The TP-Link RTL83xx based switches have their MAC address programmed
in the "para" partition. While in theory, the format of this partition
is dynamic, in practice, the MAC address appears to be located at a
consistent address. Thus, use nvmem-cells to read this MAC address.

The main MAC is required for deriving the MAC address of the switch
ports. Instead of reading it via mtd_get_mac_binary(), alias the
ethernet0 node as the label-mac-device, and use get_mac_label().

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

musashino205 commented:
Please add an instruction of OpenWrt installation to the commit message regardless of whether the method is the same for other devices.

mrnuke committed:
realtek: tl-sg2xxx: enable netowrk status LEDs

The TP-Link SG2xxx devices use an RTL8231 in shift register mode.
There is currently no kernel driver for the LED controller that drives
the RL8231. The bootloader does not configre the hardware.

The LEDs don't work in OpenWRT. Fortunately, the LED controller
registers are exported to debugfs. Use debugfs to manually initialize
the registers on boot, and enable the network status LEDs

This is a temporary solution until a kernel driver becomes available.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

svanheule commented:
> The TP-Link SG2xxx devices use an RTL8231 in shift register mode.
> There is currently no kernel driver for this mode.

And there never will be a specific driver. There are already drivers for plain shift registers, but that is not was is required here. Instead a driver for the LED peripheral is needed.

> This has not been a problem on other realtek-based switches, as the bootloader would
> configure the hardware.

People have been complaining about this though, as usually the bootloader does not set up the LEDs when cold booting a switch into OpenWrt. I have some patches that enable support for the LED peripheral: https://patchwork.ozlabs.org/project/openwrt/list/?series=309892 (but I've also been busy with upstream patches, reviewing other patches, work)

mrnuke commented:
> Instead a driver for the LED peripheral is needed.
> I have some patches that enable support for the LED peripheral: https://patchwork.ozlabs.org/project/openwrt/list/?series=309892 (but I've also been busy with upstream patches, reviewing other patches, work)

I've updated the commit message to make the distinction between LED controller and RTL8231. I can take out the script that enables the LEDs, should it be controversial. I would have to update the commit of the SG2210P to no longer say that LEDs are working, and remove instructions on how to enable the SFP.

svanheule reviewed (commented):
```
The SFP port control lines are hardwired, except for tx-disable. These
lines are controller by the RTL8231 in shift register mode. There is
no driver support for this yet.

However, to enable the lasers on SFP1 and SFP2 respectively:

    echo 0x140 > /sys/kernel/debug/rtl838x/led/led_sw_p_ctrl.26
    echo 0x140 > /sys/kernel/debug/rtl838x/led/led_sw_p_ctrl.24
```
The tx-disable lines are managed by the LED controller!? These are non-latching shift register outputs, this should be glitching horribly. Does the board have giant capacitances on these lines to "debounce" the signals?

mrnuke commented on target/linux/realtek/dts-5.10/rtl8380_tplink_sg2xxx.dtsi:
```diff
@@ -128,15 +129,25 @@
 				reg = <0x1b00000 0x400000>;
 			};
 			partition@1f00000 {
+				compatible = "nvmem-cells";
```
From my understanding from reading the flu-boot sources, the partition is split in 64KiB blocks. The last 12 bytes have a size field, ID field, and the `aa 55 aa 55` field.

The _boring way_ is to go through these blocks from high address to low address, until a block with ID of '0'. Then subtract the `size` field to get the beginning of the data. In practice I'm consistently seeing the MAC at the offset in this DTB.

mrnuke commented:
> The tx-disable lines are managed by the LED controller!? These are non-latching shift register outputs, this should be glitching horribly. Does the board have giant capacitances on these lines to "debounce" the signals?

Yes, they are controlled by the non-latching shift register. How lovely! There are two unused GPIOs on the SoC, so I have no clue why they did this.

I have some high-res images of the smaller brother if you'd like to give them a look. There are some R-C elements, so there may very well be a low-pass RC filter.

https://github.com/mrnuke/openwrt/wiki/TP-Link-TP-SG2008P

mrnuke commented on target/linux/realtek/dts-5.10/rtl8380_tplink_sg2xxx.dtsi:
```diff
@@ -0,0 +1,181 @@
+			compatible = "ti,tps23861";
+			reg = <0x30>;
+			shunt-resistor-micro-ohms = <255000>;
+		};
```
If I leave the nodes, with `status = "disabled"`, then the SG2008P gets in a bootloop. I'm not going to be debugging this:

```
[   11.250000] rtl83xx-switch switch@1b000000: configuring for fixed/internal link mode
[   11.260000] DSA: tree 0 setup
[   11.260000] LINK state irq: 20
[   11.260000] In rtl83xx_setup_qos
[   11.270000] Setting up RTL838X QoS
[   11.270000] RTL838X_PRI_SEL_TBL_CTRL(i): 00033112
[   11.280000] Current Intprio2queue setting: 00000000
[   11.280000] QM_PKT2CPU_INTPRI_MAP: 00fac688
[   11.290000] rtl838x_dbgfs_init called
[   11.300000] CPU 0 Unable to handle kernel paging request at virtual address 00000000, epc == 802c7b58, ra == 8022f8cc
[   11.310000] Oops[#1]:
[   11.310000] CPU: 0 PID: 893 Comm: kworker/0:3 Not tainted 5.10.138 #0
[   11.310000] Workqueue: events deferred_probe_work_func
[   11.310000] $ 0   : 00000000 00000001 00000000 8216cf60
[   11.310000] $ 4   : 00000000 00000000 811fd654 82084000
[   11.310000] $ 8   : 00000000 00000001 811f0000 127a3980
[   11.310000] $12   : 00000000 0000d190 00000000 00000100
[   11.310000] $16   : fffffffe 8306e440 00000000 81200000
[   11.310000] $20   : 81200000 821a4bc4 00000003 fffffffe
[   11.310000] $24   : 00000000 989c4992                  
[   11.310000] $28   : 8265a000 8265bc20 80710000 8022f8cc
[   11.310000] Hi    : 00000000
[   11.310000] Lo    : 000003c0
[   11.310000] epc   : 802c7b58 strlen+0x0/0x2c
[   11.310000] ra    : 8022f8cc start_creating.part.0+0x78/0x194
[   11.310000] Status: 1010fc03 KERNEL EXL IE 
[   11.310000] Cause : 10800008 (ExcCode 02)
[   11.310000] BadVA : 00000000
[   11.310000] PrId  : 00019070 (MIPS 4KEc)
[   11.310000] Modules linked in:
[   11.310000] Process kworker/0:3 (pid: 893, threadinfo=(ptrval), task=(ptrval), tls=00000000)
[   11.310000] Stack : 00000002 83079708 83077d60 00000000 821a4080 8306e440 00000000 00000060
[   11.310000]         00000018 802301ac 0000003c 0000000f 806a0000 80230634 83077898 00000000
[   11.310000]         00000000 806a0000 821a4080 8306e440 821a4080 803a4494 00000003 fffffffe
[   11.310000]         80710000 80230248 806a21ec 8066e0bc 821a4080 8306e440 00000018 821a4464
[   11.310000]         821a4bb8 821a4bc4 00000003 fffffffe 80710000 803a4ba0 82087700 806a13a8
[   11.310000]         ...
[   11.310000] Call Trace:
[   11.310000] [<802c7b58>] strlen+0x0/0x2c
[   11.310000] [<8022f8cc>] start_creating.part.0+0x78/0x194
[   11.310000] [<802301ac>] debugfs_create_dir+0x1c/0x15c
[   11.310000] [<803a4494>] rtl838x_dbgfs_port_init+0x54/0x258
[   11.310000] [<803a4ba0>] rtl838x_dbgfs_init+0xe0/0x56c
[   11.310000] [<807b9a7c>] rtl83xx_sw_probe+0x7d4/0x81c
[   11.310000] [<80328b7c>] platform_drv_probe+0x40/0x94
[   11.310000] [<80326448>] really_probe+0x124/0x55c
[   11.310000] [<80323d3c>] bus_for_each_drv+0x70/0xb0
[   11.310000] [<80326990>] __device_attach+0x110/0x1f0
[   11.310000] [<80325144>] bus_probe_device+0x9c/0xb8
[   11.310000] [<80325764>] deferred_probe_work_func+0xbc/0x114
[   11.310000] [<80040aa0>] process_one_work+0x1f0/0x460
[   11.310000] [<80041060>] worker_thread+0x350/0x5c4
[   11.310000] [<800473b4>] kthread+0x13c/0x144
[   11.310000] [<80001bb8>] ret_from_kernel_thread+0x14/0x1c
[   11.310000] 
[   11.310000] Code: a066ffff  1000fff7  00000000 <80820000> 10400007  00000000  00801025  80430001  1460fffe 
[   11.310000] 
[   11.570000] ---[ end trace 12da47b96c41a0b4 ]---
[   11.580000] Kernel panic - not syncing: Fatal exception
[   11.580000] Rebooting in 1 seconds..
```

mrnuke replied:
It seems that disabling the port (not phy) will cause the crash:

```
&port24 {
	status = "disabled";
};
```

I would like to model the board layout such that:
- the board is described in the dtsi
- unpopulated footprints are described in the device dts

I am going to leave the `/delete-node/` for now, unless you tell me (again) otherwise. I think this approach is the least amount of code duplication. I hope you agree.

svanheule replied:
The crash should be fixed with 9a7f17e11f5d ("realtek: ignore disabled switch ports").

If you have two 8-port devices and one 10-port devices you can support with this dtsi, then enabled those two ports by default sounds a bit backwards to me. I think using `/delete-node/` twice for sounds like more confusing than using `status = "disabled"` to indicate it's not populated.

mrnuke commented on target/linux/realtek/base-files/etc/init.d/realtek_leds:
```diff
@@ -0,0 +1,70 @@
+	echo 0x0510ff00 > "$leds"/led_p_en_ctrl
+	echo 0x00100000 > "$leds"/led0_sw_p_en_ctrl
+	echo 0x00100000 > "$leds"/led1_sw_p_en_ctrl
+	echo 0x0fffffff > "$leds"/led2_sw_p_en_ctrl
```
I am moving this commit at the end of the series, for reference. Feel free to leave it out.
(Personally, I am using this to get das blinkenlightsen)

mrnuke commented:
I have a couple of fixes that have been hidden:
- fix i2c address of tps32861
- fix port labeling

Instead of hiding them in a commit that moves the .dts, I find it clearer and more explicit to make the fixes in separate commits.

svanheule commented on target/linux/realtek/dts-5.10/rtl838x.dtsi:
```diff
@@ -45,7 +45,7 @@
 
 #define SWITCH_SFP_PORT(n, s, m) \
-	port@##n { \
+	port##n: port@##n { \
```
If you need labels, you should be able to just prefix the macros for identical results:
`port24: SWTICH_SFP_PORT(n, s, m);`

svanheule commented:
Merged the two fixes, so you'll have to rebase on master.

mrnuke committed:
realtek: rtl838x: label switch port dts nodes

Although PHY nodes are labeled, the port nodes were not. Labeling of
ports is useful for 'status = "disabled"' ports, which is supported
since commit 9a7f17e11f5d ("realtek: ignore disabled switch ports")

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
realtek: rtl8380-tl-sg2xxx: use a single "firmware" partition

The "firmware" partition was assembled from two contiguous partitions.
This complexity is unnecessary. Instead of using mtd-concat over
"sys" and "usrimg1", simply declare the "firmware" partition to cover
the flash space instead.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke commented on target/linux/realtek/dts-5.10/rtl8380_tplink_sg2xxx.dtsi:
```diff
@@ -116,13 +98,10 @@
+			partition@100000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x100000 0x1a00000>;
```
@svanheule, as promised.

svanheule replied:
Thanks!

svanheule commented:
Thanks for the updates, and for converting the T1600 to nvmem-cells for the MAC address. Except for the LED script, I've merged the patches (with amended commit messages):
bc9dcfb1ce65 realtek: split TP-Link SG2000 series devicetree
5f026f12728f realtek: rtl838x: label switch port dts nodes
d55c08739032 realtek: tl-sg2xxx: read MAC address from nvmem-cells
ab2a4c1e016e realtek: rtl8380-tl-sg2xxx: use a single "firmware" partition
01e2184c49c5 realtek: add support for TP-Link SG2210P

I'll close this PR for now. If you wish to reopen for the LED script, please bear in mind the restrictions listed here: https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr

Shine- commented:
Hi there, I just saw this PR getting merged.

Unfortunately, the commit message doesn't mention which hw version of this device this commit is for. The commit itself suggests v3.0 - which is ambiguous, see below.

Explanation: There's a myriad of versions for the SG2210P and siblings.

- TL-SG2210P v1.0
- T1500G-10PS v1.0 (also labelled TL-SG2210P v2.0 on the bottom, stock firmware not compatible with original SG2210P)
- T1500G-10PS v2.0 (also labelled TL-SG2210P v3.0 on the bottom - stock firmware is compatible with T1500G v1)
- TL-SG2210P v3.0 (NOT the same as the above two, different stock firmware, can't be crossflashed)
- TL-SG2210P v3.2 (should be same as v3.0, but again not same as T1500G v1/v2
- TL-SG2201P v3.26 (data sheet suggests this is the same as v3/v3.2)
- TL-SG2210P v4, v4.20 and v5 (at some point they started to advertise them as PoE+, not PoE anymore. Maybe different PoE chip? Or only stonger PSU??)

Same goes for the SG2008 / T1500-8G. The hw/fw versions were mostly in-sync with the SG2210P.

SG2008P, on the other hand, has different hw version numbering and firmware that doesn't look in-sync with SG2008/SG2210P/T1500G at first glance.

It should be noted in detail which version exactly your commit is for, resp. has been tested with. Especially the ambiguosity between "SG2210P v3.0" (later version) and "T1500G v2.0 (SG2210P v3.0)" (earlier version) should be resolved.

I do have multiple T1500G v1 and v2 (both 8-port non-PoE and 10-port PoE versions), but they're all in-use currently, so I can't do any testing/bricking atm, sorry.

svanheule commented:
> * T1500G-10PS v1.0 (also labelled TL-SG2210P v2.0 on the bottom, stock firmware not compatible with original SG2210P)
> 
> * T1500G-10PS v2.0 (also labelled TL-SG2210P v3.0 on the bottom - stock firmware is compatible with T1500G v1)
> 
> * TL-SG2210P v3.0 (NOT the same as the above two, different stock firmware, can't be crossflashed)


> It should be noted in detail which version exactly your commit is for, resp. has been tested with. Especially the ambiguosity between "SG2210P v3.0" (later version) and "T1500G v2.0 (SG2210P v3.0)" (earlier version) should be resolved.

If TP-Link can't even be consistent about naming their own devices, it's hard for us to "fix" that. I suspect @mrnuke has the TL-SG2210P v3.0 (non-T1500G).


> I do have multiple T1500G v1 and v2 (both 8-port non-PoE and 10-port PoE versions), but they're all in-use currently, so I can't do any testing/bricking atm, sorry.

If you get the chance to take them out of service for a few hours, opening them up to photograph the boards would also help. Otherwise we're just guessing at which devices may or may not be compatible. Different vendor firmwares doesn't have to mean the devices aren't internally very similar.

Some analysis of the FW image TP-Link provides for the devices you've listed would also be interesting. That way we would already have an idea of which devices are using Realtek hardware, and which aren't. TP-Link also provides GPL archives, which sometimes just link to the archive of another device (e.g. EAP245v3 and EAP265HD).

svanheule commented:
Image decryption isn't so much of an issue, that's something people have worked out already.
See for example:
- wiki page with image info https://svanheule.net/switches/image_formats/tplink
- @andyboeh's script to decrypt images and generate sysupgrade images from stock FW files: https://gist.github.com/andyboeh/e5cf74d2912dee527823d654414e4323

mrnuke commented:
@mrnuke has the v3.26. Whether it's v3, v3.xx, v5, or v5.6, they all point to the same `rtk-maple_gpl.tar.gz` GPL drop, indicatint RTL8380M

I wouldn't be surprised is the version bumps mean "We found a cheaper diode at the Shenzen market today".

Shine- commented:
Newer versions (starting with v4 or v5) are advertised as PoE+, not PoE, and include a stronger PSU. Who knows, perhaps there's a different PoE chip in there, perhaps not. Even my T1500G properly detect class4 PoE+ PDs. Just don't know what they'd do if the device actually requested PoE+ current levels.

As for source, T1500G GPL drop references to "rtl839x" and "maple", just like SG2210Pv3.

mrnuke commented:
> Newer versions (starting with v4 or v5) are advertised as PoE+, not PoE, and include a stronger PSU.

Are you sure?

![image](https://user-images.githubusercontent.com/2610559/189942306-01695d54-3bf0-439f-b801-5f78d0725431.png)

Shine- commented:
Haha. Other language TP-Link sites properly advertise the v1 as PoE (802.3af) with 53W and the v3 as PoE+ (802.3at/af) with 61W. As Sander said, if they're not even consistent about naming their own devices...

I'll try to take pictures of the inside of T1500G v1 and v2 one of these days. Don't have a SG2210v1, though.

svanheule commented:
@Shine-, if you notice any differences, or would like to add alternative names to the current devices, you're always welcome to submit patches :)

andyboeh commented:
Yeah, the naming scheme is a mess - that's one of the reasons why I chose TL-SG2452P instead of T1600G-52PS for my switch (both names are on the bottom sticker). IMHO it's more consistent with the other TP-Link devices I have, e.g. the TL-SG108E.

Shine- commented:
As promised, here are photos of a T1500Gv1 (SG2210Pv2). To me, this looks like exactly the same hardware, including PoE controllers (despite this version being advertised as PoE only, not PoE+).

[T1500Gv1(SG2210Pv2).zip](https://github.com/openwrt/openwrt/files/9565687/T1500Gv1.SG2210Pv2.zip)

Shine- commented:
And here the T1500Gv1(SG2008v2) - the non-PoE, but PoE-PD version of this switch. Obviously a different board, despite firmwares being in sync with the SG2210Pv2/v3.0.

[T1500Gv1(SG2008v2).zip](https://github.com/openwrt/openwrt/files/9565882/T1500Gv1.SG2008v2.zip)

mrnuke commented:
And this?

![IMG_2006](https://user-images.githubusercontent.com/2610559/190164416-74fc8f2c-7ff5-42bd-b854-f1a63eac2dbe.JPG)
![IMG_2007](https://user-images.githubusercontent.com/2610559/190164422-77d291c9-2169-4ad1-b8db-7da522b599b2.JPG)
![IMG_2008](https://user-images.githubusercontent.com/2610559/190164425-00a5ac20-b5dd-49a3-bbbe-0939adf0508f.JPG)
![IMG_2009](https://user-images.githubusercontent.com/2610559/190164426-03641d82-255a-447c-8b92-b67c9b055e01.JPG)
![IMG_2010](https://user-images.githubusercontent.com/2610559/190164428-ecb90970-d715-4dfc-8689-706e7138e51f.JPG)

Shine- commented:
Yes, exactly, if you want images displayed in line. I used a ZIP archive on purpose, but whatever.

mrnuke commented:
How many board revisions did they do? Here's a TL-SG2008P0v1 and/or TL-SG2210P-v3: https://github.com/mrnuke/openwrt/wiki/TP-Link-TP-SG2008P.

Shine- commented:
Next is the T1500G-8T v2 (SG2008v3.0). Looks pretty much identical to its predecessor.

[T1500Gv2(SG2008v3.0).zip](https://github.com/openwrt/openwrt/files/9566929/T1500Gv2.SG2008v3.0.zip)

mrnuke commented:
> v3.**0** is a T1500G-10PS. v3.**2** (3.20 or 3.26) is a SG2210P. Despite looking like exactly the same hardware, stock firmware is incompatible between those two models.

We should take this conversation elsewhere, but I'm going to request a citation on that claim, since it directly contradicts the manufacturer's support site:
![image](https://user-images.githubusercontent.com/2610559/190183886-e9d2585b-9cab-41d4-957f-8ec6dcfe03d7.png)

Shine- commented:
More photo goodness.

T1500G-10PS v2.0 and 2.1 (both also labelled SG2210P v3.0). They look completely identical to T1500Gv1.0/SG2210Pv2.0 (see photos in my comment above).

To stress it again. Stock firmware with "v3.2", "v3.20" or "v3.26" in the file name **does not** work with these v3.0 models, regardless of what the TP-Link support site claims. IMHO, the model supported by **this PR** should be modified to the one that @mrnuke actually **tested it** on: v3.26 (or at least v3.2), until someone can actually confirm that it works with earlier versions, like 2.0, 2.1 or 3.0.

[T1500Gv2.0(SG2210Pv3.0).zip](https://github.com/openwrt/openwrt/files/9582986/T1500Gv2.0.SG2210Pv3.0.zip)
[T1500Gv2.1(SG2210Pv3.0).zip](https://github.com/openwrt/openwrt/files/9582990/T1500Gv2.1.SG2210Pv3.0.zip)

Shine- commented:
And now to something more interesting: TL-SG2210P v5.0

This one looks like a slightly different design, I assume a cost-saving mod. The RTL8231 is missing and it features different PoE controllers.

[SG2210Pv5.0.zip](https://github.com/openwrt/openwrt/files/9583054/SG2210Pv5.0.zip)

svanheule commented:
> And now to something more interesting: TL-SG2210P v5.0 [...] features different PoE controllers.

[MP3924](https://www.monolithicpower.com/en/mp3924.html) PSE chips. The vendor provides a freely available and detailed datasheet, how nice!

mrnuke commented:
@Shine- , thank you for sharing the images. Notice how the TL-SG2008P/TL-SG2210P-v3.26 (left) has The reset IC **U5**, replaced with a new chip, **U37**, when compared to T1500Gv2.0(SG2210Pv3.0). U37 also has a watchdog timer, that, I believe U5 lacks.

From my experience, the bootloader sets a GPIO to disable this watchdog. I am not sure if the kernel actually uses the watchdog, or leaves it disabled. There's also a couple of extra populated resistors on the v3.0 near the JTAG GPIOs, that are not installed on v3.26

![image](https://user-images.githubusercontent.com/2610559/190884400-0e0648c2-e897-4699-93cc-83fc4489aab5.png)

Shine- commented:
More photos, this time a TL-SG2008v1. Significant differences to later HW versions is that it's can't be PoE-PD powered and has smaller flash and RAM.

[SG2008v1.zip](https://github.com/openwrt/openwrt/files/9614093/SG2008v1.zip)

This concludes the different HW versions of this switch series, that I have access to.

Shine- commented:
P.S.: Just in case anyone's wondering. You are allowed, even encouraged to re-use and distribute the photos I posted here. You can use them for any purpose and without any restrictions, not even credit required. All the photos are my own personal work, nobody else was involved in creating them.

damex commented:
> To stress it again. Stock firmware with "v3.2", "v3.20" or "v3.26" in the file name **does not** work with these v3.0 models, regardless of what the TP-Link support site claims. IMHO, the model supported by **this PR** should be modified to the one that @mrnuke actually **tested it** on: v3.26 (or at least v3.2), until someone can actually confirm that it works with earlier versions, like 2.0 or 3.0.

i have just had an actual v3.0 and v3.20 revisions of sg2210p to test. neither of them was able to boot what this PR built saying about 'bag magic header'. v3.20 revision had winbond chip and it is very sensitive. shoring it to ground first time worked but shorting it to second time killed poor nor chip (well, needed to try to run it once again). v3.0 had the same winbond chip and survived few shorts so far

pinout on both is TX/RX/GND/3.3v counting from |> pointing to TX symbol next to j2 marking on board. nothing says about it on this PR. TX resistor traces could be bridged straight to TX pin for ease.

- shorting chips is not a good idea. especially bad idea making end user do it and risk their device in process
- making end user solder extra tiny resistors on tiny traces is a bad idea
- PR did not worked on hardware it claims that it should work

mrnuke commented:
> > To stress it again. Stock firmware with "v3.2", "v3.20" or "v3.26" in the file name **does not** work with these v3.0 models, regardless of what the TP-Link support site claims. IMHO, the model supported by **this PR** should be modified to the one that @mrnuke actually **tested it** on: v3.26 (or at least v3.2), until someone can actually confirm that it works with earlier versions, like 2.0 or 3.0.
> 
> i have just had an actual v3.0 and v3.20 revisions of sg2210p to test. neither of them was able to boot what this PR built saying about 'bag magic header'. v3.20 revision had winbond chip and it is very sensitive. shoring it to ground first time worked but shorting it to second time killed poor nor chip (well, needed to try to run it once again). v3.0 had the same winbond chip and survived few shorts so far

I suspect an electrostatic discharge may have been involved. The #RESET is an input pin to the chip, so you can't damage the ROM simply by shorting its input to ground. The circuit that drives #RESET is also protected by a 50 Ohm resistor, so it's unlikely to damage that either. Here's the simplified schematic:

https://github.com/mrnuke/openwrt/wiki/TP-Link-TP-SG2008P#reset-and-watchdog-circuit


>     * shorting chips is not a good idea. especially bad idea making end user do it and risk their device in process
>     * making end user solder extra tiny resistors on tiny traces is a bad idea

What I can do is document what it takes to get OpenWRT installed. It's up to the user to take this information and make a decision on the amount of risk they are willing to tolerate. I can not make that decision for anyone.

>     * PR did not worked on hardware it claims that it should work

Might be best to open up an OpenWRT forum thread on this. What did the serial console say? What's the vendor's update file look like?

Shine- commented:
Hi guys, I have some news to tell. I got my hands on a TL-SG2210P v3.0 (ie. without dual-labelling as T1500G). So this device does, in fact, exist. It's advertised as PoE+ on the retail box and label (not just PoE like its T1500G counterpart) and it's using the v3.2 firmware, so it's firmware-incompatible to the T1500G labelled devices. It also has the changed reset circuit as outlined by @mrnuke above, so it appears to be identical hardware to v3.2/v3.26, but different hardware than the dual-labelled T1500Gv2/SG2210Pv3.0.

Please find photos attached.
[SG2210Pv3.0.zip](https://github.com/openwrt/openwrt/files/10479647/SG2210Pv3.0.zip)

Shine- commented:
Just FYI, in case anyone's interested: Apparently, since stock firmware version dated 2022-11-30 and onwards, TP-Link unified the firmware, so the stock firmware files for TL-SG2210P v3[.2x], v4 and v5 are binary identical, despite the differences in hardware. Even this unified firmware isn't installabe on the dual-labelled TL-SG2210P v3 / T1500G-10PS v2.x, though.
